collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java

```diff
@@ -0,0 +1,708 @@
+
+                throw new Exception(error);
+            }
+        }
+
+        InputStream downloadStream = entity.getContent();
+
+        String hash = "";
+
+        if (HTTP_CONTENT_TYPE_TEXT_XML.equals(contentType)) {
+            byte[] bytes = IOUtils.toByteArray(downloadStream);
+            downloadStream = new ByteArrayInputStream(bytes);
+            hash = FileUtils.getMd5Hash(new ByteArrayInputStream(bytes));
+        }
+
+        Header contentEncoding = entity.getContentEncoding();
+        if (contentEncoding != null && contentEncoding.getValue().equalsIgnoreCase(GZIP_CONTENT_ENCODING)) {
+            downloadStream = new GZIPInputStream(downloadStream);
+        }
+
+        Map<String, String> responseHeaders = new HashMap<>();
+
+        Header[] fields = response.getAllHeaders();
+
+        if (fields != null && fields.length >= 1) {
+            for (Header h : fields) {
+                responseHeaders.put(h.getName(), h.getValue());
+            }
+        }
+
+        return new HttpGetResult(downloadStream, responseHeaders, hash, statusCode);
+    }
+
+
+    @Override
+    public HttpHeadResult head(@NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws Exception {
+        addCredentialsForHost(uri, credentials);
+        getCookieStore().clear();
+
+        HttpContext localContext = getHttpContext();
+        HttpClient httpclient = createHttpClient(CONNECTION_TIMEOUT);
+        HttpHead httpHead = createOpenRosaHttpHead(uri);
+        Map<String, String> responseHeaders = new HashMap<>();
+
+        // if https then enable preemptive basic auth...
+        if (uri.getScheme() != null && uri.getScheme().equals("https")) {
+            enablePreemptiveBasicAuth(localContext, uri.getHost());
+        }
+
+        final HttpResponse response;
+        int statusCode;
+
+        try {
+            Timber.i("Issuing HEAD request to: %s", uri.toString());
+
+            response = httpclient.execute(httpHead, localContext);
+            statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode == HttpStatus.SC_UNAUTHORIZED) {
+                getCookieStore().clear();
+
+            } else if (statusCode == HttpStatus.SC_NO_CONTENT) {
+                for (Header head : response.getAllHeaders()) {
+                    responseHeaders.put(head.getName(), head.getValue());
+                }
+            }
+
+            discardEntityBytes(response);
+
+        } catch (ClientProtocolException | ConnectTimeoutException | UnknownHostException | SocketTimeoutException | NoHttpResponseException | SocketException e) {
+            String errorMessage;
+
+            if (e instanceof ClientProtocolException) {
+                errorMessage = "Client Protocol Exception";
+            } else if (e instanceof ConnectTimeoutException) {
+                errorMessage = "Connection Timeout";
+            } else if (e instanceof UnknownHostException) {
```

Comment:
I'm not sure I like this whole exception block. Not sure if this is legacy code or what but I think setting the `errorMessage` to the name of the exception is a bit redundant.

Reply:
I agree it's not the cleanest - It's kind of legacy with a few mods.. I'll take a look, but again HttpClientConnection will be deleted and replaced with another one we are happy with the refactor.